web3.js/src/loader.js

```diff
@@ -45,36 +45,87 @@ export class Loader {
    * @param program Account to load the program into
    * @param programId Public key that identifies the loader
    * @param data Program octets
+   * @return true if program was loaded successfully, false if program was already loaded
    */
   static async load(
     connection: Connection,
     payer: Account,
     program: Account,
     programId: PublicKey,
     data: Buffer | Uint8Array | Array<number>,
-  ): Promise<void> {
+  ): Promise<boolean> {
     {
       const balanceNeeded = await connection.getMinimumBalanceForRentExemption(
         data.length,
       );
-      const transaction = new Transaction().add(
-        SystemProgram.createAccount({
-          fromPubkey: payer.publicKey,
-          newAccountPubkey: program.publicKey,
-          lamports: balanceNeeded > 0 ? balanceNeeded : 1,
-          space: data.length,
-          programId,
-        }),
-      );
-      await sendAndConfirmTransaction(
-        connection,
-        transaction,
-        [payer, program],
-        {
-          commitment: 'single',
-          skipPreflight: true,
-        },
+
+      // Fetch program account info to check if it has already been created
+      const programInfo = await connection.getAccountInfo(
+        program.publicKey,
+        'single',
       );
+
+      let transaction: Transaction | null = null;
+      if (programInfo !== null) {
+        if (programInfo.executable) {
+          return false;
+        }
+
+        if (programInfo.data.length !== data.length) {
```

Comment:
Allocate will only allocate an empty account, maybe check against zero?

Reply:
Yeah definitely true. I initially approached this with more exhaustive checks to ensure an invalid transaction wouldn't be submitted. But ultimately felt that relying on transaction simulation would be nicer and simplify the client implementation. Right now we are skipping the preflight check but I have an upcoming PR that will send all transactions through with preflight check enabled. How do you feel about that?

Reply:
@jackcmay any thoughts here?

Reply:
Are you thinking of getting rid of all these checks and relying on preflight only?  Keeping the exists, executable, and owner checks would be helpful.  Also, probably don't have to do preflight for the `Write`s (or maybe just do it for the first one)

Reply:
Was thinking all of the checks in this PR. I think of them more as put the account in the correct state if at all possible, not trying to prevent transaction errors. 

> Also, probably don't have to do preflight for the Writes (or maybe just do it for the first one)

Is this because simulation would slow things down? The downside of skipping preflight for later writes is that we wouldn't catch "out of funds" errors... the transactions would just silently be dropped

Reply:
Think we can handle the necessary funds calculations on the client side?

Reply:
Yeah, sounds good. I can take care of that along with #12261 then and skip preflight for all but the first write